File: app/src/main/java/com/texnar13/deliveryapp/MainViewModel.java
package com.texnar13.deliveryapp;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.texnar13.deliveryapp.model.DBAddress;
import com.texnar13.deliveryapp.model.DBExpedition;
import com.texnar13.deliveryapp.model.DBNotification;
import com.texnar13.deliveryapp.model.DBPackage;
import com.texnar13.deliveryapp.model.DBUser;

import org.bson.Document;
import org.bson.conversions.Bson;
import org.bson.types.ObjectId;

import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

import io.realm.mongodb.App;
import io.realm.mongodb.AppConfiguration;
import io.realm.mongodb.Credentials;
import io.realm.mongodb.User;
import io.realm.mongodb.mongo.MongoClient;
import io.realm.mongodb.mongo.MongoCollection;
import io.realm.mongodb.mongo.MongoDatabase;
import io.realm.mongodb.mongo.iterable.MongoCursor;

public class MainViewModel extends ViewModel {

    private static final String MY_LOG = "Test";

    private String apiKey;
    private MongoDatabase mongoDatabase = null;

// ------------------------------------------ Связь с активностью ---------------------------------------

    // строка для отправки тостов
    public MutableLiveData<String> toastMessage = new MutableLiveData<>();

    private void sendToast(String message) {
        toastMessage.setValue(message);
    }

    // статус подключения
    enum ConnectionStatusValue {
        STATUS_NONE,
        STATUS_ERROR,
        STATUS_CONNECTED,
    }

    private final MutableLiveData<ConnectionStatusValue> connectionStatus =
            new MutableLiveData<>(ConnectionStatusValue.STATUS_NONE);
    public final LiveData<ConnectionStatusValue> activityConnectionStatus = connectionStatus;


// ------------------------------------------ Загруженные данные ------------------------------------------

    // текущий пользователь
    public MutableLiveData<DBUser> currentUser = new MutableLiveData<>();

    // уведомления
    public MutableLiveData<List<DBNotification>> currentUserNotifications = new MutableLiveData<>();

    // посылки пользователя
    public MutableLiveData<List<DBExpedition>> currentUserExpeditions = new MutableLiveData<>();


// ------------------------------------------ Главные методы ---------------------------------------

// todo это соответственно переносится во ViewModel,
//            //   а enableLoadBar() работает через подписку на ViewModel
//            //   нажатие кнопок и обратня связь от фрагментов делегируется во viewModel
//            //   LiveData и MutableLiveData :)
//            //   Можно сделать так, чтобы LiveData следила за MutableLiveData и избежать getter-ов и setter-ов
//            //   контекст view model не хранит, но он передаётся в методах
//
//            //


    public MainViewModel(String apiKey) {
        this.apiKey = apiKey;
        init();
    }


    // нициализация
    private void init() {
        Log.e("Hello", "VM init");

        // подключение к онлайн бд
        connectDB();
    }

    // уничтожение активности
    @Override
    protected void onCleared() {
        Log.e("Hello", "VM destroy");
        super.onCleared();
    }


// ---------------------------------------------- БД -----------------------------------------------

    // подключение к онлайн бд
    void connectDB() {
        String appID = "application-0-epiwn";
        App realmConnectionApp = new App(new AppConfiguration.Builder(appID).build());
        // авторизация
        Credentials apiKeyCredentials = Credentials.apiKey(apiKey);
        AtomicReference<User> user = new AtomicReference<>();
        realmConnectionApp.loginAsync(apiKeyCredentials, it -> {
            if (it.isSuccess()) {
                Log.e(MY_LOG, "Successfully authenticated using an API Key.");
                user.set(realmConnectionApp.currentUser());

                MongoClient mongoClient = realmConnectionApp.currentUser().getMongoClient("mongodb-atlas");
                // получаем ссылку на БД
                mongoDatabase = mongoClient.getDatabase("DeliveryApplicationDB");

                // уведомляем активность, что подключение установлено
                connectionStatus.setValue(ConnectionStatusValue.STATUS_CONNECTED);
            } else {
                Log.e(MY_LOG, it.getError().toString());

                // уведомляем активность, что подключение не произошло
                connectionStatus.setValue(ConnectionStatusValue.STATUS_ERROR);

                // Повторная попытка
                connectionStatus.setValue(ConnectionStatusValue.STATUS_NONE);
                // отложенный вызов
                (new Handler(Looper.getMainLooper())).postDelayed(this::connectDB, 4000);
            }
        });

    }

    // выход из учетной записи пользователя
    void logout() {
        currentUser.setValue(null);
    }


    void tryRegisterUser(String password, String[] address, String email, String name, String phone) {

        // получаем таблицу пользователей
        MongoCollection<Document> usersCollection = mongoDatabase.getCollection(DBUser.TABLE_NAME);

        // поиск пользователя в бд по email
        Document query = new Document(DBUser.USER_EMAIL, email);
        usersCollection.find(query).first().getAsync(result -> {
            if (result.isSuccess()) {
                Document existingUser = result.get();

                // если пользователь найден
                if (existingUser != null) {
                    sendToast("Пользователь с таким E-mail уже существует");
                } else {
                    // создаем пользователя

                    // создаем новый документ пользователя
                    Document newUser = new Document()
                            .append(DBUser.USER_PASSWORD, hashPassword(password))
                            .append(DBUser.USER_ADDRESS,
                                    DBAddress.getDocumentFromAddressArray(address))
                            .append(DBUser.USER_EMAIL, email)
                            .append(DBUser.USER_NAME, name)
                            .append(DBUser.USER_PHONE_NUMBER, phone)
                            .append(DBUser.USER_PICTURE, "URL")
                            .append(DBUser.USER_RATING, 0d);

                    // вставка нового пользователя в коллекцию
                    usersCollection.insertOne(newUser).getAsync(insertResult -> {
                        if (insertResult.isSuccess()) {
                            sendToast("Пользователь успешно создан");

                            // и сразу загружаем и заходим за этого пользователя
                            authUser(email, password);
                        } else {
                            sendToast("Ошибка при создании пользователя");
                        }
                    });
                }
            } else {
                sendToast("Ошибка при проверке существования пользователя");
            }
        });
    }


    void authUser(String email, String password) {

        if (mongoDatabase == null) return;

        // получаем таблицу пользователей
        MongoCollection<Document> usersCollection = mongoDatabase.getCollection(DBUser.TABLE_NAME);

        // поиск пользователя в бд
        Document query = new Document(DBUser.USER_EMAIL, email);
        usersCollection.find(query).first().getAsync(result -> {
            if (result.isSuccess()) {
                Document documentUser = result.get();

                // если пользователь найден
                if (documentUser != null) {
                    String pass = documentUser.getString(DBUser.USER_PASSWORD);

                    if (hashPassword(password).equals(pass)) {
                        // вытягиваем данные из пользователя и сохраняем их
                        currentUser.setValue(new DBUser(documentUser));
                        sendToast("Авторизован");

                        // Получаем данные пользователя
                        loadUserNotifications();
                        loadUserExpeditions();
                    } else {
                        sendToast("Пароль неподходит!");
                    }
                } else
                    sendToast("Пользователь не найден");
            } else {
                sendToast("Ошибка подключения");
            }
        });
    }

    // редактирование пользователя
    public void editUser(@NonNull DBUser editedUserData) {
        Document editedDocument = editedUserData.getDocument();

        // получаем таблицу пользователей
        MongoCollection<Document> usersCollection = mongoDatabase.getCollection(DBUser.TABLE_NAME);

        // Получаем идентификатор пользователя из измененных данных
        ObjectId userId = editedDocument.getObjectId(DBUser.USER_ID);

        // Обновляем документ пользователя новыми данными
        usersCollection.findOneAndReplace(new Document(DBUser.USER_ID, userId), editedDocument).getAsync(result -> {
            if (result.isSuccess()) {
                Document updatedUser = result.get();
                if (updatedUser != null) {
                    // если данные сохранены успешно, обновляем глобальную копию переменной и интерфейс
                    currentUser.setValue(editedUserData);
                    sendToast("Данные пользователя успешно сохранены");
                } else {
                    sendToast("Ошибка, пользователь не найден");
                }
            } else {
                sendToast("Ошибка при сохранении данных пользователя");
            }
        });
    }

    void loadUserNotifications() {
        DBUser user = currentUser.getValue();
        if (user != null) {

            // получаем таблицу уведомлений
            MongoCollection<Document> notificationsCollection = mongoDatabase.getCollection(DBNotification.TABLE_NAME);

            // поиск непрочитанных уведомлений пользователя в бд
            Document query = new Document(DBNotification.NOTIFICATION_USER, user.get_id()).append(
                    DBNotification.NOTIFICATION_STATUS, DBNotification.NOTIFICATION_STATUS_UNREAD);

            notificationsCollection.find(query).iterator().getAsync(result -> {
                if (result.isSuccess()) {
                    // сохраняем все уведомления в лист
                    MongoCursor<Document> cursor = result.get();

                    // пробегаемся по всем уведомлениям
                    List<DBNotification> loadedNotifications = new LinkedList<>();
                    while (cursor.hasNext())
                        loadedNotifications.add(new DBNotification(cursor.next()));

                    // передаем получившийся лист в глобальный отслеживаемый
                    currentUserNotifications.setValue(loadedNotifications);
                }
            });
        }
    }

    // пометить уведомление прочитанным
    public void markReadAdminNotification(ObjectId notificationID) {

        // получаем таблицу уведомлений
        MongoCollection<Document> notificationsCollection = mongoDatabase.getCollection(DBNotification.TABLE_NAME);

        // Создаем запрос для поиска документа по идентификатору уведомления
        Document query = new Document(DBNotification.NOTIFICATION_ID, notificationID);

        // Обновление поля
        Document update = new Document("$set", new Document(
                DBNotification.NOTIFICATION_STATUS, DBNotification.NOTIFICATION_STATUS_BEEN_READ));

        // Выполнение обновления
        notificationsCollection.updateOne(query, update).getAsync(result -> {
            if (result.isSuccess()) {
                loadUserNotifications();
            } else {
                sendToast("Ошибка при обновлении поля");
            }
        });
    }

    // загрузить отправления пользователя
    void loadUserExpeditions() {
        DBUser user = currentUser.getValue();
        if (user != null) {

            // получаем таблицу отправлений
            MongoCollection<Document> expeditionsCollection = mongoDatabase.getCollection(DBExpedition.TABLE_NAME);

            // поиск отправлений пользователя в бд
            Document query = new Document(DBExpedition.EXPEDITION_SENDER, user.get_id());

            expeditionsCollection.find(query).iterator().getAsync(result -> {
                if (result.isSuccess()) {

                    // сохраняем все отправления в лист
                    MongoCursor<Document> cursor = result.get();

                    // пробегаемся по всем посылкам
                    List<DBExpedition> loadedData = new ArrayList<>();
                    while (cursor.hasNext())
                        loadedData.add(new DBExpedition(cursor.next()));

                    // передаем получившийся лист в глобальный отслеживаемый
                    currentUserExpeditions.setValue(loadedData);
                }
            });
        }
    }

    public void createExpedition(DBExpedition expedition) {
        DBUser user = currentUser.getValue();
        if (user != null) {

            // получаем таблицу отправлений
            MongoCollection<Document> expeditionsCollection = mongoDatabase.getCollection(DBExpedition.TABLE_NAME);

            Document expeditionDocument = new Document()
                    .append(DBExpedition.EXPEDITION_ADDRESS_RECEIVER, expedition.getAddressReceiver().getDocument())
                    .append(DBExpedition.EXPEDITION_ADDRESS_SENDER, expedition.getAddressSender().getDocument())
                    .append(DBExpedition.EXPEDITION_STATUS, expedition.getStatus())
                    .append(DBExpedition.EXPEDITION_SENDER, user.get_id())
                    .append(DBExpedition.EXPEDITION_PACKAGE, expedition.getPackage().toDocument());

            // вставка нового пользователя в коллекцию
            expeditionsCollection.insertOne(expeditionDocument).getAsync(insertResult -> {
                if (insertResult.isSuccess()) {
                    sendToast("Отправление успешно создано");

                    // и сразу загрузка всего заново
                    loadUserExpeditions();

                } else {
                    sendToast("Ошибка при создании отправления = " + insertResult.getError());
                    Log.e("Test", "Ошибка при создании отправления = " + insertResult.getError());
                }
            });

        }
    }

    public void editExpedition(DBExpedition editedExpeditionData) {
        DBUser user = currentUser.getValue();
        if (user != null) {

            // получаем таблицу отправлений
            MongoCollection<Document> expeditionsCollection = mongoDatabase.getCollection(DBExpedition.TABLE_NAME);

            // документ редактирования
            Document editedDocument = editedExpeditionData.getDocument();


            // Обновляем документ пользователя новыми данными
            expeditionsCollection.findOneAndReplace(
                    new Document(DBExpedition.EXPEDITION_ID, editedExpeditionData.get_id()),
                    editedDocument
            ).getAsync(result -> {
                if (result.isSuccess()) {
                    Document updatedDocument = result.get();
                    if (updatedDocument != null) {
                        // если данные сохранены успешно, обновляем глобальную копию переменной и интерфейс
                        loadUserExpeditions();

                        sendToast("Данные отправления успешно сохранены");
                    } else {
                        sendToast("Ошибка, отправление не найдено");
                    }
                } else {
                    sendToast("Ошибка при сохранении данных отправления");
                }
            });

        }
    }


// -------------------------------------------------------------------------------------------------

    public static String hashPassword(String password) {
        try {
            // Создание экземпляра MessageDigest с алгоритмом SHA-256
            MessageDigest md = MessageDigest.getInstance("SHA-256");

            // Преобразование пароля в массив байтов и хэширование
            byte[] hashedBytes = md.digest(password.getBytes());

            // Преобразование массива байтов в шестнадцатеричную строку
            BigInteger number = new BigInteger(1, hashedBytes);
            StringBuilder hexString = new StringBuilder(number.toString(16));

            // Обрезка строки до 16 символов для простоты
            String shortHash = hexString.substring(0, 16);

            return shortHash;
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

}

File: app/src/main/java/com/texnar13/deliveryapp/view_model/MainViewModel.kt
package com.texnar13.deliveryapp.view_model

import android.util.Log
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.texnar13.deliveryapp.model.entities.EntityAddress
import com.texnar13.deliveryapp.model.entities.EntityExpedition
import com.texnar13.deliveryapp.model.entities.EntityExpedition.Companion.ExpeditionStatus
import com.texnar13.deliveryapp.model.entities.EntityPackage
import com.texnar13.deliveryapp.model.entities.EntityTrip
import com.texnar13.deliveryapp.model.entities.EntityUser
import com.texnar13.deliveryapp.model.http.HttpApi
import com.texnar13.deliveryapp.model.shared_preferences.SPHolder
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.math.BigInteger
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.Date


class MainViewModel(
    private val httpClient: HttpApi,
    private val spHolder: SPHolder
) : ViewModel(), HttpApi.HttpResultAndStatusListener {


// ------------------------------------------- Константы -------------------------------------------

    companion object {

        private const val TAG = "VIEW_MODEL"


        // получение экземпляра ViewModel из фрагмента
        fun getViewModel(activity: FragmentActivity): MainViewModel {
            return ViewModelProvider(
                activity,
                MainViewModelFactory(activity)
            )[MainViewModel::class.java]
        }

        fun hashPassword(password: String): String {
            try {
                // Создание экземпляра MessageDigest с алгоритмом SHA-256
                val md = MessageDigest.getInstance("SHA-256")

                // Преобразование пароля в массив байтов и хэширование
                val hashedBytes = md.digest(password.toByteArray())

                // Преобразование массива байтов в шестнадцатеричную строку
                val number = BigInteger(1, hashedBytes)
                val hexString = StringBuilder(number.toString(16))

                // Обрезка строки до 16 символов для простоты
                val shortHash = hexString.substring(0, 16)

                return shortHash
            } catch (e: NoSuchAlgorithmException) {
                throw RuntimeException(e)
            }
        }

    }


// -------------------------------------------------------------------------------------------------
// -------------------------------------- Загруженные данные ---------------------------------------
// -------------------------------------------------------------------------------------------------


    var token: MutableLiveData<String?> = MutableLiveData(null)

    // текущий пользователь
    var currentUser: MutableLiveData<EntityUser?> = MutableLiveData(null)

    // посылки пользователя
    val currentUserExpeditions: MutableLiveData<List<EntityExpedition>?> = MutableLiveData(null)

    val selectedExpedition: MutableLiveData<EntityExpedition?> = MutableLiveData(null)


    // загруженные поездки
    var currentLoadedTrips: MutableLiveData<List<EntityTrip>> = MutableLiveData()


// -------------------------------------------------------------------------------------------------
// ------------------------------------------ Конструктор ------------------------------------------
// -------------------------------------------------------------------------------------------------

    // нициализация
    init {
        Log.i(TAG, "init")

        // назначаем слушатель HTTP ответов и статуса
        httpClient.setHttpResultListener(this)
    }


// -------------------------------------------------------------------------------------------------
// ------------------------------------- Отправка уведомлений --------------------------------------
// -------------------------------------------------------------------------------------------------

    // строка для отправки тостов
    var toastMessage: MutableLiveData<String> = MutableLiveData()

    private fun sendToast(message: String) {
        toastMessage.postValue(message)
    }

// -------------------------------------------------------------------------------------------------
// ------------------------------------- Прогресс бар загрузки -------------------------------------
// -------------------------------------------------------------------------------------------------

    private val mutableHttpLoadingStatus =
        MutableLiveData(HttpApi.Companion.HttpClientState.NO_WORK)
    val httpLoadingStatus: LiveData<HttpApi.Companion.HttpClientState> = mutableHttpLoadingStatus

    // Статус HTTP интерфейса
    override fun onHttpStatusUpdated(state: HttpApi.Companion.HttpClientState) {

        viewModelScope.launch {
            // чтобы разметка корректно отрабатывала события
            delay(100)
            mutableHttpLoadingStatus.postValue(state)
        }
    }


// -------------------------------------------------------------------------------------------------
// ----------------------------------- Назначение адреса сервера -----------------------------------
// -------------------------------------------------------------------------------------------------

    fun getServerAddress(): String {
        return spHolder.getServerAddress()
    }

    fun setServerAddress(address: String) {
        spHolder.setServerAddress(address)
        httpClient.updateServerAddress(address)
        sendToast("Сохранено")
    }

// -------------------------------------------------------------------------------------------------
// -------------------------- Регистрация пользователя (Создание учётки) ---------------------------
// -------------------------------------------------------------------------------------------------

    fun tryRegisterUser(
        email: String,
        password: String,
        address: EntityAddress,//[4]
        name: String,
        phone: String
    ) {
        Log.i(TAG, "tryRegisterUser")

        // в любом случае сохраняем последние введённые поля
        spHolder.setUserLastAuth(
            email,
            password
        )

        viewModelScope.launch {

            // пытаемся аутентифицировать пользователя
            httpClient.createNewUser(
                email = email,
                password = password,
                address = address,
                name = name,
                phone = phone
            )
        }
    }

    // обратная связь от Api
    override fun httpCreateUserFailure(errorCode: HttpApi.Companion.ErrorCode, status: String) {
        Log.i(TAG, "httpCreateUserFailure status=$status")
        sendToast("Ошибка: $status")
    }

    // обратная связь от Api
    override fun httpCreateUserSuccess(token: String) {
        Log.i(TAG, "httpCreateUserSuccess token=$token")
        sendToast("Пользователь успешно создан. Загрузка профиля")

        // сохраняем созданый токен
        this.token.postValue(token)

        viewModelScope.launch {
            // вытягиваем данные из пользователя и сохраняем их
            httpClient.getUserData(token)
        }
    }


// -------------------------------------------------------------------------------------------------
// --------------------------------- Аутентификация пользователя -----------------------------------
// -------------------------------------------------------------------------------------------------

    // получение уже введённых ранее логина и пароля
    fun getUserLastAuth(): Array<String> {
        return spHolder.getUserLastAuth()
    }

    fun authUser(email: String, password: String) {
        Log.i(TAG, "authUser email=$email password=$password")

        // отправляем асинхроннный запрос
        viewModelScope.launch {
            // вход пользователя
            httpClient.loginUser(email, password)
        }

        // в любом случае сохраняем последние введённые поля
        spHolder.setUserLastAuth(
            email,
            password
        )
    }

    // обратная связь от Api
    override fun httpLoginUserFailure(errorCode: HttpApi.Companion.ErrorCode, status: String) {
        Log.i(TAG, "httpLoginUserFailure status=$status")
        sendToast("Ошибка: $status")
    }

    // обратная связь от Api
    override fun httpLoginUserSuccess(token: String) {
        Log.i(TAG, "httpLoginUserSuccess token=$token")
        // сохраняем полученый токен
        this.token.postValue(token)

        sendToast("Авторизация успешна. Загрузка профиля")

        // отправляем асинхроннный запрос
        viewModelScope.launch {
            // вытягиваем данные из пользователя и сохраняем их
            httpClient.getUserData(token)
        }
    }

    // Получение данных пользователя неуспешно
    override fun httpGetUserDataFailure(errorCode: HttpApi.Companion.ErrorCode, status: String) {
        Log.i(TAG, "httpGetUserDataFailure status=$status")
        sendToast("Ошибка: $status")

    }

    // Данные пользователя получены
    override fun httpGetUserDataSuccess(user: EntityUser) {
        Log.i(TAG, "httpGetUserDataSuccess")
        sendToast("Данные загружены")
        currentUser.postValue(user)
    }


// -------------------------------------------------------------------------------------------------
// --------------------------------- Страничка пользователя -----------------------------------
// -------------------------------------------------------------------------------------------------


    // выход из учетной записи пользователя
    fun logout() {
        Log.i(TAG, "logout")
        token.postValue(null)
    }

    // редактирование пользователя
    private var tempEditedUserData: EntityUser? = null
    fun editUser(editedUserData: EntityUser) {
        tempEditedUserData = editedUserData

        val token = token.value
        if (token != null) {
            // отправляем асинхроннный запрос
            viewModelScope.launch {
                httpClient.editUserData(token, editedUserData)
            }
        }
    }

    override fun httpEditUserDataFailure(errorCode: HttpApi.Companion.ErrorCode, status: String) {
        when (errorCode) {
            HttpApi.Companion.ErrorCode.TOKEN_EXPIRED -> {
                sendToast("Сессия истекла: $status")
                // Выходим из текущего пользователя
                logout()
            }

            else -> sendToast("Ошибка при сохранении: $status")
        }
    }

    override fun httpEditUserDataSuccess() {
        sendToast("Данные пользователя успешно сохранены")
        currentUser.postValue(tempEditedUserData)
    }


// -------------------------------------------------------------------------------------------------
// -------------------------------------- Страничка маршрутов --------------------------------------
// -------------------------------------------------------------------------------------------------


    fun loadTrips(
        departCountry:String,
        departCity: String,
        destinationCountry:String,
        destinationCity: String,
        departDate: Date,
        freeWeightMin: Float
    ) {// todo


        val token = token.value
        if (token != null) {
            viewModelScope.launch {
                httpClient.loadTrajectoriesByParams(
                    token,
                    departCountry,
                    departCity,
                    destinationCountry,
                    destinationCity,
                    departDate,
                    freeWeightMin
                )
            }
        }
        //        // получаем таблицу
//        MongoCollection<Document> tripsCollection = mongoDatabase.getCollection(DBTrip.TABLE_NAME);
//
//        // получаем последние 10
//        tripsCollection.find().limit(10).iterator().getAsync(result -> {
//            if (result.isSuccess()) {
//                // сохраняем все в лист
//                MongoCursor<Document> cursor = result.get();
//
//                // пробегаемся по всем
//                List<DBTrip> loadedNotifications = new LinkedList<>();
//                while (cursor.hasNext())
//                    loadedNotifications.add(new DBTrip(cursor.next()));
//
//                // передаем получившийся лист в глобальный отслеживаемый
//                currentLoadedTrips.setValue(loadedNotifications);
//            }
//        });
    }


    override fun httpLoadTrajectoriesFailure(
        errorCode: HttpApi.Companion.ErrorCode,
        status: String
    ) {
        when (errorCode) {
            HttpApi.Companion.ErrorCode.TOKEN_EXPIRED -> {
                sendToast("Сессия истекла: $status")
                // Выходим из текущего пользователя
                logout()
            }

            else -> sendToast("Ошибка: $status")
        }
    }

    override fun httpLoadTrajectoriesSuccess(trips: List<EntityTrip>) {
        currentLoadedTrips.postValue(trips)
    }


// -------------------------------------------------------------------------------------------------
// ------------------------------------- Страничка отправлений -------------------------------------
// -------------------------------------------------------------------------------------------------


    // загрузить отправления пользователя
    fun loadUserExpeditions() {
        // обнуляем
        currentUserExpeditions.postValue(null)

        val token = token.value
        if (token != null) {
            // Загружаем отправления пользователя
            viewModelScope.launch {
                httpClient.loadUserExpeditions(token)
            }
        }
    }

    // ответ Http загрузчика
    override fun httpLoadUserExpeditionsFailure(
        errorCode: HttpApi.Companion.ErrorCode,
        status: String
    ) {
        when (errorCode) {
            HttpApi.Companion.ErrorCode.TOKEN_EXPIRED -> {
                sendToast("Сессия истекла: $status")
                // Выходим из текущего пользователя
                logout()
            }

            else -> sendToast("Ошибка: $status")
        }
    }

    // ответ Http загрузчика
    override fun httpLoadUserExpeditionsSuccess(expeditions: List<EntityExpedition>) {
        currentUserExpeditions.postValue(expeditions)

    }

    // создание нового отправления
    fun createExpedition(
        id: Long,
        addressReceiver: EntityAddress,
        addressSender: EntityAddress,
        status: ExpeditionStatus,
        expeditionPackage: EntityPackage
    ) {
        val expedition = EntityExpedition(
            id,
            addressReceiver,
            addressSender,
            status,
            currentUser.value!!.id,
            null,
            expeditionPackage
        )

        val token = token.value
        if (token != null) {
            // создание нового отправления
            viewModelScope.launch {
                httpClient.createUserExpedition(token, expedition)
            }
        }
    }


    // ответ создания
    override fun httpCreateExpeditionFailure(
        errorCode: HttpApi.Companion.ErrorCode,
        status: String
    ) {
        when (errorCode) {
            HttpApi.Companion.ErrorCode.TOKEN_EXPIRED -> {
                sendToast("Сессия истекла")
                // Выходим из текущего пользователя
                logout()
            }

            else ->
                sendToast("Ошибка при создании отправления = $status")
        }
    }

    // ответ создания
    override fun httpCreateExpeditionSuccess(expedition: EntityExpedition) {

        sendToast("Отправление успешно создано")

        loadUserExpeditions()
    }

    // выбираем из разметки отправление которое будет редактироваться (для вывода в диалог)
    // может передаваться null тогда это будет диалог создания
    fun selectExpeditionForEdit(editedExpeditionData: EntityExpedition?) {

        selectedExpedition.value = editedExpeditionData
    }

    // Редактируем отправление
    fun editExpedition(
        addressReceiver: EntityAddress,
        addressSender: EntityAddress,
        status: ExpeditionStatus,
        expeditionPackage: EntityPackage
        ) {
        val editedExpeditionData = EntityExpedition(
            selectedExpedition.value!!.id,
            addressReceiver,
            addressSender,
            status,
            selectedExpedition.value!!.senderId,
            selectedExpedition.value!!.courierId,
            expeditionPackage
        )

        val token = token.value
        if (token != null) {
            // Редактируем отправление
            viewModelScope.launch {
                httpClient.editExpedition(token, editedExpeditionData)
            }
        }
    }

    // ответ редактирования
    override fun httpEditExpeditionFailure(errorCode: HttpApi.Companion.ErrorCode, status: String) {
        when (errorCode) {
            HttpApi.Companion.ErrorCode.TOKEN_EXPIRED -> {
                sendToast("Сессия истекла")
                // Выходим из текущего пользователя
                logout()
            }

            else ->
                sendToast("Ошибка при редактировании отправления = $status")
        }
    }

    // ответ редактирования
    override fun httpEditExpeditionSuccess(expedition: EntityExpedition) {
        sendToast("Данные сохранены")
        loadUserExpeditions()
    }


    val trajectoryAndExpeditionData = MutableLiveData<EntityTrip?>(null)

    // Для функции "посмотреть маршрут"
    fun loadTrajectoryDataForExpedition(expedition: EntityExpedition) {
        val token = token.value
        if (token != null) {
            // Редактируем отправление
            viewModelScope.launch {
                httpClient.loadTrajectoryDataForExpedition(token, expedition)
            }
        }
    }

    // ответ "посмотреть маршрут"
    override fun loadTrajectoryDataForExpeditionFailure(
        errorCode: HttpApi.Companion.ErrorCode,
        status: String
    ) {
        TODO("Not yet implemented")
    }

    // ответ "посмотреть маршрут"
    override fun loadTrajectoryDataForExpeditionSuccess(trajectoryAndExpedition: EntityTrip) {
        trajectoryAndExpeditionData.postValue(trajectoryAndExpedition)
    }

//--------------------------------------------------------------

    val selectedTrip = MutableLiveData<EntityTrip?>(null)

    // "Сделать заявку" сопрячь
    fun selectTripForEdit(tripUnit: EntityTrip) {

        // Ставим поездку выбранной
        selectedTrip.value = tripUnit

        // Запускаем загрузку отправлений пользователя
        loadUserExpeditions()

    }

    // в "Сделать заявку" была выбрана посылка
    fun selectPackageDeliveryTrip(expedition: EntityExpedition) {
        val token = token.value
        if (token != null && selectedTrip.value != null) {
            // Редактируем отправление
            viewModelScope.launch {
                httpClient.selectPackageDeliveryTrip(token, expedition, selectedTrip.value!!)
            }
        }

    }

    override fun selectPackageDeliveryTripFailure() {
        TODO("Not yet implemented")
    }

    override fun selectPackageDeliveryTripSuccess() {
        sendToast("API говорит ДА...")
    }
}

